app/integration_action.go

```diff
@@ -0,0 +1,189 @@
+	}
+
+	request := &model.PostActionIntegrationRequest{
+		UserId:    userId,
+		ChannelId: post.ChannelId,
+		TeamId:    channel.TeamId,
+		PostId:    postId,
+		Type:      action.Type,
+		Context:   action.Integration.Context,
+	}
+
+	clientTriggerId, _, err := request.GenerateTriggerId(a.AsymmetricSigningKey())
+	if err != nil {
+		return "", err
+	}
+
+	if action.Type == model.POST_ACTION_TYPE_SELECT {
+		request.DataSource = action.DataSource
+		request.Context["selected_option"] = selectedOption
+	}
+
+	resp, err := a.DoActionRequest(action.Integration.URL, request.ToJson())
+	if resp != nil {
+		defer consumeAndClose(resp)
+	}
+	if err != nil {
+		return "", err
+	}
+
+	var response model.PostActionIntegrationResponse
+	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+		return "", model.NewAppError("DoPostAction", "api.post.do_action.action_integration.app_error", nil, "err="+err.Error(), http.StatusBadRequest)
+	}
+
+	retainedProps := []string{"override_username", "override_icon_url"}
+
+	if response.Update != nil {
+		response.Update.Id = postId
+		response.Update.AddProp("from_webhook", "true")
+		for _, prop := range retainedProps {
+			if value, ok := post.Props[prop]; ok {
+				response.Update.Props[prop] = value
+			} else {
+				delete(response.Update.Props, prop)
+			}
+		}
+		if _, err := a.UpdatePost(response.Update, false); err != nil {
+			return "", err
+		}
+	}
+
+	if response.EphemeralText != "" {
+		ephemeralPost := &model.Post{}
+		ephemeralPost.Message = model.ParseSlackLinksToMarkdown(response.EphemeralText)
+		ephemeralPost.ChannelId = post.ChannelId
+		ephemeralPost.RootId = post.RootId
+		if ephemeralPost.RootId == "" {
+			ephemeralPost.RootId = post.Id
+		}
+		ephemeralPost.UserId = post.UserId
+		ephemeralPost.AddProp("from_webhook", "true")
+		for _, prop := range retainedProps {
+			if value, ok := post.Props[prop]; ok {
+				ephemeralPost.Props[prop] = value
+			} else {
+				delete(ephemeralPost.Props, prop)
+			}
+		}
+		a.SendEphemeralPost(userId, ephemeralPost)
+	}
+
+	return clientTriggerId, nil
+}
+
+// Perform an HTTP POST request to an integration's action endpoint.
+// Caller must consume and close returned http.Response as necessary.
+func (a *App) DoActionRequest(rawURL string, body []byte) (*http.Response, *model.AppError) {
+	req, _ := http.NewRequest("POST", rawURL, bytes.NewReader(body))
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+
+	// Allow access to plugin routes for action buttons
+	var httpClient *httpservice.Client
+	url, _ := url.Parse(rawURL)
+	siteURL, _ := url.Parse(*a.Config().ServiceSettings.SiteURL)
+	subpath, _ := utils.GetSubpathFromConfig(a.Config())
+	if (url.Hostname() == "localhost" || url.Hostname() == "127.0.0.1" || url.Hostname() == siteURL.Hostname()) && strings.HasPrefix(url.Path, path.Join(subpath, "plugins")) {
```

Comment:
Can we add this functionality to the httpservice rather than special casing it here?

Reply:
This isn't code I added I just moved it around. And are we sure want don't want to special case it? Is there any case where we wouldn't want to allow this?

Reply:
Ya. I don't see a reason why we wouldn't want to always allow it.

Reply:
I agree, will update it

Reply:
Actually now that I start looking at the http service code more closely it's a lot more refactoring to move this in there than I want to tackle in this PR. So unless you're 5/5 I'm going to defer to take care of it in some future refactoring since this isn't code I wrote